File: core/src/main/java/com/emarsys/core/worker/CoreCompletionHandlerMiddleware.java
package com.emarsys.core.worker;

import android.os.Handler;

import com.emarsys.core.CoreCompletionHandler;
import com.emarsys.core.database.repository.Repository;
import com.emarsys.core.database.repository.SqlSpecification;
import com.emarsys.core.handler.CoreSdkHandler;
import com.emarsys.core.request.factory.DefaultRunnableFactory;
import com.emarsys.core.request.factory.RunnableFactory;
import com.emarsys.core.request.model.RequestModel;
import com.emarsys.core.request.model.RequestModelKt;
import com.emarsys.core.request.model.specification.FilterByRequestIds;
import com.emarsys.core.response.ResponseModel;
import com.emarsys.core.util.Assert;
import com.emarsys.core.util.RequestModelUtils;

import java.util.Arrays;

public class CoreCompletionHandlerMiddleware implements CoreCompletionHandler {
    CoreCompletionHandler coreCompletionHandler;
    Repository<RequestModel, SqlSpecification> requestRepository;
    Worker worker;
    CoreSdkHandler coreSDKHandler;
    Handler uiHandler;
    RunnableFactory runnableFactory;

    public CoreCompletionHandlerMiddleware(
            Worker worker,
            Repository<RequestModel, SqlSpecification> requestRepository,
            Handler uiHandler,
            CoreSdkHandler coreSDKHandler,
            CoreCompletionHandler coreCompletionHandler) {
        Assert.notNull(requestRepository, "RequestRepository must not be null!");
        Assert.notNull(worker, "Worker must not be null!");
        Assert.notNull(coreCompletionHandler, "CoreCompletionHandler must not be null!");
        Assert.notNull(uiHandler, "uiHandler must not be null!");
        Assert.notNull(coreSDKHandler, "coreSDKHandler must not be null!");
        this.coreCompletionHandler = coreCompletionHandler;
        this.requestRepository = requestRepository;
        this.worker = worker;
        this.coreSDKHandler = coreSDKHandler;
        this.runnableFactory = new DefaultRunnableFactory();
        this.uiHandler = uiHandler;
    }

    @Override
    public void onSuccess(final String id, final ResponseModel responseModel) {
        coreSDKHandler.post(runnableFactory.runnableFrom(new Runnable() {
            @Override
            public void run() {
                removeRequestModel(responseModel);

                worker.unlock();
                worker.run();

                handleSuccess(responseModel);
            }
        }));
    }

    @Override
    public void onError(final String id, final ResponseModel responseModel) {
        coreSDKHandler.post(runnableFactory.runnableFrom(new Runnable() {
            @Override
            public void run() {
                if (isNonRetriableError(responseModel.getStatusCode())) {
                    removeRequestModel(responseModel);

                    handleError(responseModel);
                    worker.unlock();
                    worker.run();
                } else {
                    worker.unlock();
                }
            }
        }));
    }

    private void removeRequestModel(ResponseModel responseModel) {
        String[] ids = RequestModelKt.collectRequestIds(responseModel.getRequestModel());

        int noOfIterations = ids.length % 500 == 0 ? ids.length / 500 : ids.length / 500 + 1;
        for (int i = 0; i < noOfIterations; i++) {
            int noOfElements = Math.min(ids.length, (i + 1) * 500);
            requestRepository.remove(new FilterByRequestIds(Arrays.copyOfRange(ids, i * 500, noOfElements)));
        }
    }

    @Override
    public void onError(final String id, final Exception cause) {
        coreSDKHandler.post(runnableFactory.runnableFrom(new Runnable() {
            @Override
            public void run() {
                worker.unlock();
                uiHandler.post(runnableFactory.runnableFrom(new Runnable() {
                    @Override
                    public void run() {
                        coreCompletionHandler.onError(id, cause);
                    }
                }));
            }
        }));
    }

    private boolean isNonRetriableError(int statusCode) {
        if (statusCode == 408 || statusCode == 429) {
            return false;
        } else {
            return 400 <= statusCode && statusCode < 500;
        }
    }

    private void handleSuccess(final ResponseModel responseModel) {
        for (final String id : RequestModelUtils.extractIdsFromCompositeRequestModel(responseModel.getRequestModel())) {
            uiHandler.post(runnableFactory.runnableFrom(new Runnable() {
                @Override
                public void run() {
                    coreCompletionHandler.onSuccess(id, responseModel);
                }
            }));
        }
    }

    private void handleError(final ResponseModel responseModel) {
        for (final String id : RequestModelUtils.extractIdsFromCompositeRequestModel(responseModel.getRequestModel())) {
            uiHandler.post(runnableFactory.runnableFrom(new Runnable() {
                @Override
                public void run() {
                    coreCompletionHandler.onError(id, responseModel);
                }
            }));
        }
    }
}

File: core/src/main/java/com/emarsys/core/worker/CoreCompletionHandlerMiddleware.kt
package com.emarsys.core.worker

import com.emarsys.core.CoreCompletionHandler
import com.emarsys.core.Mockable
import com.emarsys.core.database.repository.Repository
import com.emarsys.core.database.repository.SqlSpecification
import com.emarsys.core.handler.ConcurrentHandlerHolder
import com.emarsys.core.request.model.RequestModel
import com.emarsys.core.request.model.collectRequestIds
import com.emarsys.core.request.model.specification.FilterByRequestIds
import com.emarsys.core.response.ResponseModel
import com.emarsys.core.util.RequestModelUtils
import kotlin.math.min

@Mockable
class CoreCompletionHandlerMiddleware(
    private var worker: Worker?,
    private val requestRepository: Repository<RequestModel, SqlSpecification>,
    var concurrentHandlerHolder: ConcurrentHandlerHolder,
    val coreCompletionHandler: CoreCompletionHandler?
) : CoreCompletionHandler {

    override fun onSuccess(id: String, responseModel: ResponseModel) {
        concurrentHandlerHolder.post {
            removeRequestModel(responseModel)
            worker?.unlock()
            worker?.run()
            handleSuccess(responseModel)
        }
    }

    override fun onError(id: String, responseModel: ResponseModel) {
        concurrentHandlerHolder.post {
            if (isNonRetriableError(responseModel.statusCode)) {
                removeRequestModel(responseModel)
                handleError(responseModel)
                worker?.unlock()
                worker?.run()
            } else {
                worker?.unlock()
            }
        }
    }

    private fun removeRequestModel(responseModel: ResponseModel) {
        val ids = responseModel.requestModel.collectRequestIds()
        val maximumRequestCountInTransaction = 50
        val noOfIterations =
            if (ids.size % maximumRequestCountInTransaction == 0) ids.size / maximumRequestCountInTransaction else ids.size / maximumRequestCountInTransaction + 1
        for (i in 0 until noOfIterations) {
            val noOfElements = min(ids.size, (i + 1) * maximumRequestCountInTransaction)
            requestRepository.remove(
                FilterByRequestIds(
                    ids.copyOfRange(i * maximumRequestCountInTransaction, noOfElements)
                )
            )
        }
    }

    override fun onError(id: String, cause: Exception) {
        concurrentHandlerHolder.post {
            worker?.unlock()
            concurrentHandlerHolder.postOnMain {
                coreCompletionHandler?.onError(
                    id,
                    cause
                )
            }
        }
    }

    private fun isNonRetriableError(statusCode: Int): Boolean {
        return if (statusCode == 408 || statusCode == 429) {
            false
        } else {
            400 <= statusCode && statusCode < 500
        }
    }

    private fun handleSuccess(responseModel: ResponseModel) {
        for (id in RequestModelUtils.extractIdsFromCompositeRequestModel(responseModel.requestModel)) {
            concurrentHandlerHolder.postOnMain {
                coreCompletionHandler?.onSuccess(
                    id,
                    responseModel
                )
            }
        }
    }

    private fun handleError(responseModel: ResponseModel) {
        for (id in RequestModelUtils.extractIdsFromCompositeRequestModel(responseModel.requestModel)) {
            concurrentHandlerHolder.postOnMain {
                coreCompletionHandler?.onError(
                    id,
                    responseModel
                )
            }
        }
    }
}